### ISSUE TITLE
Multiple crashes leading into cv_wait_common->schedule

### ISSUE THREAD

**Issue by sempervictus** [zfsonlinux/spl#538]
ZFS #4464 and others lead into this call as they crash, and now getting these on send/recv too, starting to think the almighty hates me:

```
[  720.386007] INFO: task zfs:12149 blocked for more than 180 seconds.
[  720.386658]       Tainted: P           OE   4.4.3-ssv-i7 #ssv
[  720.387259] "echo 0 > /proc/sys/kernel/hung_task_timeout_secs" disables this message.
[  720.388067] zfs             D ffff880ad87fb888     0 12149  12147 0x00000000
[  720.388072]  ffff880ad87fb888 ffff880ad85d2000 ffff880ad85d4000 ffff880ad87fc000
[  720.388075]  ffff880ad87fbb18 ffff880ad87fbb58 ffff880ad87fbb40 0000000000000000
[  720.388077]  ffff880ad87fb8a0 ffffffff827b9955 ffff880ad87fbb50 ffff880ad87fb900
[  720.388080] Call Trace:
[  720.388090]  [<ffffffff827b9955>] schedule+0x35/0x80
[  720.388102]  [<ffffffffc00f221d>] cv_wait_common+0xed/0x120 [spl]
[  720.388107]  [<ffffffff820bc5f0>] ? prepare_to_wait_event+0xf0/0xf0
[  720.388115]  [<ffffffffc00f2265>] __cv_wait+0x15/0x20 [spl]
[  720.388160]  [<ffffffffc044e02a>] dmu_recv_stream+0x62a/0xaa0 [zfs]
[  720.388207]  [<ffffffffc04c1a59>] zfs_ioc_recv+0x209/0xbe0 [zfs]
[  720.388211]  [<ffffffff820a1be2>] ? default_wake_function+0x12/0x20
[  720.388213]  [<ffffffff820bbec5>] ? __wake_up_common+0x55/0x90
[  720.388222]  [<ffffffffc00f3002>] ? tsd_set+0x142/0x430 [spl]
[  720.388249]  [<ffffffffc043b12e>] ? dmu_buf_rele+0xe/0x10 [zfs]
[  720.388289]  [<ffffffffc0478a40>] ? rrw_exit+0x60/0x150 [zfs]
[  720.388294]  [<ffffffff821d4448>] ? __kmalloc+0x58/0x1a0
[  720.388300]  [<ffffffffc00eae8c>] ? strdup+0x3c/0x60 [spl]
[  720.388345]  [<ffffffffc04bf9b8>] zfsdev_ioctl+0x458/0x490 [zfs]
[  720.388348]  [<ffffffff821f2de1>] ? new_sync_read+0x81/0xa0
[  720.388352]  [<ffffffff82205aad>] do_vfs_ioctl+0x2cd/0x4b0
[  720.388358]  [<ffffffff82061c47>] ? __do_page_fault+0x1b7/0x430
[  720.388360]  [<ffffffff82205d09>] SyS_ioctl+0x79/0x90
[  720.388364]  [<ffffffff827bd3f6>] entry_SYSCALL_64_fastpath+0x16/0x75
[  720.388368] INFO: task send_traverse:12494 blocked for more than 180 seconds.
[  720.389117]       Tainted: P           OE   4.4.3-ssv-i7 #ssv
[  720.389722] "echo 0 > /proc/sys/kernel/hung_task_timeout_secs" disables this message.
[  720.390550] send_traverse   D ffff880497527538     0 12494      2 0x00000000
[  720.390554]  ffff880497527538 ffff880c241c8000 ffff880067036000 ffff880497528000
[  720.390558]  ffff880ad87f7c20 ffff880ad87f7c60 ffff880ad87f7c48 0000000000000000
[  720.390564]  ffff880497527550 ffffffff827b9955 ffff880ad87f7c58 ffff8804975275b0
[  720.390567] Call Trace:
[  720.390575]  [<ffffffff827b9955>] schedule+0x35/0x80
[  720.390589]  [<ffffffffc00f221d>] cv_wait_common+0xed/0x120 [spl]
[  720.390593]  [<ffffffff820bc5f0>] ? prepare_to_wait_event+0xf0/0xf0
[  720.390604]  [<ffffffffc00f2265>] __cv_wait+0x15/0x20 [spl]
[  720.390650]  [<ffffffffc043f0a2>] bqueue_enqueue+0x62/0xe0 [zfs]
[  720.390705]  [<ffffffffc044ae53>] send_cb+0x153/0x170 [zfs]
[  720.390758]  [<ffffffffc044ee43>] traverse_visitbp+0x4c3/0x870 [zfs]
[  720.390775]  [<ffffffffc00eaf74>] ? spl_kmem_alloc+0xc4/0x170 [spl]
[  720.390820]  [<ffffffffc044ec9f>] traverse_visitbp+0x31f/0x870 [zfs]
[  720.390872]  [<ffffffffc044ec9f>] traverse_visitbp+0x31f/0x870 [zfs]
[  720.390911]  [<ffffffffc044ec9f>] traverse_visitbp+0x31f/0x870 [zfs]
[  720.390950]  [<ffffffffc044ec9f>] traverse_visitbp+0x31f/0x870 [zfs]
[  720.390997]  [<ffffffffc044f270>] traverse_dnode+0x80/0x180 [zfs]
[  720.391030]  [<ffffffffc044f005>] traverse_visitbp+0x685/0x870 [zfs]
[  720.391062]  [<ffffffffc044ec9f>] traverse_visitbp+0x31f/0x870 [zfs]
[  720.391094]  [<ffffffffc044ec9f>] traverse_visitbp+0x31f/0x870 [zfs]
[  720.391126]  [<ffffffffc044ec9f>] traverse_visitbp+0x31f/0x870 [zfs]
[  720.391158]  [<ffffffffc044ec9f>] traverse_visitbp+0x31f/0x870 [zfs]
[  720.391190]  [<ffffffffc044ec9f>] traverse_visitbp+0x31f/0x870 [zfs]
[  720.391222]  [<ffffffffc044ec9f>] traverse_visitbp+0x31f/0x870 [zfs]
[  720.391253]  [<ffffffffc044f270>] traverse_dnode+0x80/0x180 [zfs]
[  720.391285]  [<ffffffffc044f0e8>] traverse_visitbp+0x768/0x870 [zfs]
[  720.391317]  [<ffffffffc044f370>] ? traverse_dnode+0x180/0x180 [zfs]
[  720.391349]  [<ffffffffc044f7ef>] traverse_impl+0x18f/0x3e0 [zfs]
[  720.391353]  [<ffffffff820a6d22>] ? sched_clock_cpu+0x72/0xa0
[  720.391385]  [<ffffffffc0449a50>] ? byteswap_record+0x2a0/0x2a0 [zfs]
[  720.391416]  [<ffffffffc044fa99>] traverse_dataset+0x59/0x60 [zfs]
[  720.391448]  [<ffffffffc044ad00>] ? dmu_recv_end_check+0x200/0x200 [zfs]
[  720.391479]  [<ffffffffc0449a85>] send_traverse_thread+0x35/0x80 [zfs]
[  720.391487]  [<ffffffffc00ed2f1>] thread_generic_wrapper+0x71/0x80 [spl]
[  720.391495]  [<ffffffffc00ed280>] ? __thread_exit+0x20/0x20 [spl]
[  720.391499]  [<ffffffff82096f69>] kthread+0xc9/0xe0
[  720.391502]  [<ffffffff82096ea0>] ? kthread_park+0x60/0x60
[  720.391506]  [<ffffffff827bd78f>] ret_from_fork+0x3f/0x70
[  720.391509]  [<ffffffff82096ea0>] ? kthread_park+0x60/0x60
```

This time in send/recv on the same host (using znapzend to send/recv every 15m).
Ping @dweeezil and @ryao: is this in the right place or should i stick this in the ZFS tracker too?

**Comment by dweeezil** [comment]
@sempervictus I suspect this belongs in the zfs tracker.  The work in zfsonlinux/zfs@fcff0f35 made some rather big changes to the manner in which send/receive work.  I didn't pay much attention to that patch when it was committed and didn't realize how extensive of a modification it actually is.  It does revert from current master code with only a few minor conflicts if you felt like testing whether it's causing the problem.

**Comment by sempervictus** [comment]
Unfortunately I can't test anymore because that pool is a single disk local
backup tier and couldn't stay broken, I murdered and rebuilt the single
disk pool... It now works just fine.

Reason I put this here to begin with is that all roads lead to Rome, and
all stack traces lead here... What's the right way fwd? Create a new zfs
issue linking this or transfer the whole issue despite not being able to
test any fixes on the same pool?

**Comment by dweeezil** [comment]
@sempervictus I finally got around to looking at the details of zfsonlinux/zfs@fcff0f3 in more detail and it appears your sender is likely blocked because it's queued 16MiB (controlled by ```zfs_send_queue_length``` which probably ought to be made a tunable in ZoL) of data which means the actual cause of the blockage may be that the receiver isn't consuming any more.  Did you see any stalled tasks on the receiving side?  Could there have been some blockage in the connection between the sender and receiver?

**Comment by behlendorf** [comment]
Closing.  Resolved by zfsonlinux/zfs@3b0d99289a8e473b059605a83d4ce0bd505ec3d3.
